### PR TITLE
Added a check for default values in question variants

### PR DESCRIPTION
**Pull request by berroar**
Amended the '_ensure_relevant_variant_fields_are_consistent' validation method to include a check that asserts that default values are the same across question variants.

### CODE REVIEW

File: app/validation/validator.py [hunk 1]
```diff
@@ -243,10 +249,15 @@ def _ensure_relevant_variant_fields_are_consistent(self, block, variants):
             errors.append(self._error_message(
                 'Variants have more than one question type for block: {}. Found types: {}'.format(block['id'], question_types)))
 
+        if len(default_answers) != 1:
```
Comment (JamesGardiner) [inline app/validation/validator.py]:
```suggestion
        if len(default_answers) > 1:
```

File: app/validation/validator.py [hunk 2]
```diff
@@ -234,6 +236,10 @@ def _ensure_relevant_variant_fields_are_consistent(self, block, variants):
             for answer in question_variant['answers']:
                 answer_ids.add(answer['id'])
                 answer_types[answer['id']].add(answer['type'])
+                try:
```
Comment (JamesGardiner) [inline app/validation/validator.py]:
```Python
default_answers.add(answer['default'], 'no default')
```

**Comment by JamesGardiner** [comment]
[Don't store default answers in the answer store (S)](https://trello.com/c/uQGZoGnS/3058-dont-store-default-answers-in-the-answer-store-s)

File: app/validation/validator.py [hunk 3]
```diff
@@ -1207,17 +1206,24 @@ def _validate_placeholder_answer_ids(self, block_id, placeholder_definition, ans
...
                 errors.append(self._error_message('Invalid answer id reference `{}` for placeholder `{}` (self-reference)'.format(
                     answer_id_to_validate,
                     placeholder_definition['placeholder']
                 )))
+                continue
+
+            non_mandatory_error = self._validate_placeholders_reference_mandatory_answers(
+                answers_with_parent_ids[answer_id_to_validate]['answer'])
+            if non_mandatory_error:
+                errors.extend(non_mandatory_error)
```
Comment (ajmaddaford) [inline app/validation/validator.py]: I don't think the extra method is needed. Also the answer id and placeholder name should be in the error message:

```python
answer = answers_with_parent_ids[answer_id_to_validate]['answer'] 
if not answer.get('mandatory'):
    errors.append(self._error_message('Placeholder references a non-mandatory answer `{}` for placeholder `{}`'.format(
        answer_id_to_validate,
        placeholder_definition['placeholder']
    )))
```

Comment (JamesGardiner) [reply app/validation/validator.py]: Are you referring to ` _validate_placeholders_reference_mandatory_answers`?

File: app/validation/validator.py [hunk 4]
```diff
@@ -207,41 +206,41 @@ def _ensure_relevant_variant_fields_are_consistent(self, block, variants):
...
-        for variant in variants:
-            question_variant = variant['question']
-            question_ids.add(question_variant['id'])
-            question_types.add(question_variant['type'])
-
-            number_of_answers.add(len(variant['question']['answers']))
-
-            if len(number_of_answers) > 1:
-                errors.append(self._error_message('Variants in block: {} contain different numbers of answers'.format(block['id'])))
+        self._get_question_variant_fields_sets(variants, question_ids, question_types,
```
[the remark this answers is not in the document]
Comment (JamesGardiner) [reply app/validation/validator.py]: Agreed :+1:

**Pull request merged**